sdks/python/apache_beam/runners/common.py

```diff
@@ -51,117 +51,192 @@ def receive(self, windowed_value):
+        len(defaults) arguments in args.
     """
-    self.step_name = step_name
-    self.window_fn = windowing.windowfn
-    self.tagged_receivers = tagged_receivers
-    self.scoped_metrics_container = (scoped_metrics_container
-                                     or ScopedMetricsContainer())
+    self.args = args
+    self.defaults = defaults
+    self._method_value = method_value
 
-    global_window = GlobalWindow()
+  def call(self, input_args, input_kwargs):
+    """Invokes the method represented by this object.
 
-    # Need to support multiple iterations.
-    side_inputs = list(side_inputs)
+    Args:
+      input_args: a list of arguments to be passed when invoking the method.
+      input_kwargs: a dictionary of keyword arguments to be passed when invoking
+                    the method.
 
-    if logging_context:
-      self.logging_context = logging_context
-    else:
-      self.logging_context = get_logging_context(logger, step_name=step_name)
+    Using a regular method call here instead of __call__ to reduce per-element
+    overhead.
+    """
+    return self._method_value(*input_args, **input_kwargs)
 
-    # Optimize for the common case.
-    self.main_receivers = as_receiver(tagged_receivers[None])
 
-    # TODO(sourabh): Deprecate the use of context
-    if state:
-      assert context is None
-      self.context = DoFnContext(self.step_name, state=state)
+class DoFnSignature(object):
+  """Represents the signature of a given ``DoFn`` object.
+
+  Signature of a ``DoFn`` provides a view of the properties of a given ``DoFn``.
+  Among other things, this will give an extensible way for for (1) accessing the
+  structure of the ``DoFn`` including methods and method parameters
+  (2) identifying features that a given ``DoFn`` support, for example, whether
+  a given ``DoFn`` is a Splittable ``DoFn`` (
+  https://s.apache.org/splittable-do-fn) (3) validating a ``DoFn`` based on the
+  feature set offered by it.
+  """
+
+  def __init__(self, do_fn):
+    # We add a property here for all methods defined by Beam DoFn features.
+    self.process_method = None
+    self.start_bundle_method = None
+    self.finish_bundle_method = None
+
+    assert isinstance(do_fn, core.DoFn)
+    self.do_fn = do_fn
+
+    def _create_do_fn_method(do_fn, method_name):
+      arguments, _, _, defaults = do_fn.get_function_arguments(method_name)
+      defaults = defaults if defaults else []
+      method_value = getattr(do_fn, method_name)
+      return DoFnMethodWrapper(method_value, arguments, defaults)
+
+    self.process_method = _create_do_fn_method(do_fn, 'process')
+    self.start_bundle_method = _create_do_fn_method(do_fn, 'start_bundle')
+    self.finish_bundle_method = _create_do_fn_method(do_fn, 'finish_bundle')
+    self._validate()
+
+  def _validate(self):
+    # start_bundle and finish_bundle methods should only have ContextParam as a
+    # default argument.
+    self._validate_start_bundle()
+    self._validate_finish_bundle()
```

Comment:
can you add a noop --> self._validate_process()

Also I wonder if this should be part of the DoFn/ParDo constructor instead of here so that we can throw creation time errors instead of runtime. I can move them around in a separate PR to keep this focussed.